cmd/zrepl/zrepl.c

```diff
@@ -207,6 +207,10 @@ uzfs_zvol_io_receiver(void *arg)
 			}
 		}
 
+		if (zinfo->state == ZVOL_INFO_STATE_OFFLINE) {
```

Comment:
How do you know that the state will not be changed to offline just after checking this cond? The other thread will wait in a loop with 1 second sleep interval waiting for this one to release its reference. How is it better than previous code then?

Reply:
This is better in sense that we are not taking lock(avoid system call of taking mutex) by using atomic_inc and atomic_dec.

There is no gurantee that state will not change to offline after this check, infact we can not add state check after every few line. If state become offline after this check then also this IO will go disk because dataset is still intact and very much available for use.

When it will receive next IO, it will come to know that state has changed and it will drop refcount and exit. Now you may argue that what if after state change, there is no IO on uzfs_zvol_io_receiver ....in that case it will stuck forever ....but that is not the case. If you look into uzfs_mark_offline_and_free_zinfo(), it change state of zinfo to offline then send a signal to uzfs_zvol_io_ack_sender() thread, very first thing uzfs_zvol_io_ack_sender thread do after wake up is to check state... so it will see state offline...will exit by closing fd. Since  uzfs_zvol_io_receiver is also waiting on read on this fd, it will observe event on fd and it will error out after fd close.

I hope things got little clear.

Reply:
thanks for explaining. If I understand correctly these checks for offline state are not necessary because if zinfo is about to be destroyed, io ack thread will close fd and the read syscalls will fail hence this thread would exit regardless.